core/src/main/scala/kafka/log/Log.scala

```diff
@@ -723,16 +728,15 @@ class Log(@volatile var dir: File,
         }
 
         // maybe roll the log if this segment is full
-        val segment = maybeRoll(messagesSize = validRecords.sizeInBytes,
-          maxTimestampInMessages = appendInfo.maxTimestamp,
-          maxOffsetInMessages = appendInfo.lastOffset)
+        val segment = maybeRoll(validRecords.sizeInBytes,
+          appendInfo)
 
         val logOffsetMetadata = LogOffsetMetadata(
-          messageOffset = appendInfo.firstOffset,
+          messageOffset = appendInfo.firstOrLastOffset,
           segmentBaseOffset = segment.baseOffset,
           relativePositionInSegment = segment.size)
 
-        segment.append(firstOffset = appendInfo.firstOffset,
+        segment.append(firstOffset = appendInfo.firstOrLastOffset,
```

Comment:
As far as I can tell, the first offset here is used only for insertion into the offset index and for logging. I cannot think of a good reason why we should prefer to use the first offset over the last offset. Since the replicas have to use the last offset for the old message format anyway, I wonder if we should try to be consistent. It would also make the logging less confusing.

Reply:
Done